plugins/influxdb/influxdb.go

```diff
@@ -0,0 +1,142 @@
+package influxdb
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/Sirupsen/logrus"
+	"github.com/stripe/veneur/plugins"
+	"github.com/stripe/veneur/samplers"
+)
+
+var _ plugins.Plugin = &InfluxDBPlugin{}
+
+// InfluxDBPlugin is a plugin for emitting metrics to InfluxDB.
+type InfluxDBPlugin struct {
+	Logger     *logrus.Logger
+	InfluxURL  string
+	HTTPClient *http.Client
+	Statsd     *statsd.Client
+}
+
+// NewInfluxDBPlugin creates a new Influx Plugin.
+func NewInfluxDBPlugin(logger *logrus.Logger, addr string, consistency string, db string, client *http.Client, stats *statsd.Client) *InfluxDBPlugin {
+	plugin := &InfluxDBPlugin{
+		Logger:     logger,
+		HTTPClient: client,
+		Statsd:     stats,
+	}
+
+	inurl, err := url.Parse(addr)
+	if err != nil {
+		logger.Fatalf("Error parsing URL for InfluxDB: %q", err)
+	}
+
+	// Construct a path we will be using later.
+	inurl.Path = "/write"
+	q := inurl.Query()
+	q.Set("db", db)
+	q.Set("precision", "s")
+	inurl.RawQuery = q.Encode()
+	plugin.InfluxURL = inurl.String()
+
+	return plugin
+}
+
+// Flush sends a slice of metrics to InfluxDB
+func (p *InfluxDBPlugin) Flush(metrics []samplers.DDMetric, hostname string) error {
+	p.Statsd.Gauge("flush.post_metrics_total", float64(len(metrics)), nil, 1.0)
+	// Check to see if we have anything to do
+	if len(metrics) == 0 {
+		p.Logger.Info("Nothing to flush, skipping.")
+		return nil
+	}
+
+	buff := bytes.Buffer{}
+	colons := regexp.MustCompile(":")
+	for _, metric := range metrics {
+		tags := strings.Join(metric.Tags, ",")
+		cleanTags := colons.ReplaceAllLiteralString(tags, "=")
+		buff.Write(
+			[]byte(fmt.Sprintf("%s,%s value=%f %d\n", metric.Name, cleanTags, metric.Value[0][1], int64(metric.Value[0][0]))),
+		)
+	}
+
+	p.postHelper(p.InfluxURL, &buff)
+
+	return nil
+}
+
+// Name returns the name of the plugin.
+func (p *InfluxDBPlugin) Name() string {
+	return "influxdb"
+}
+
+// Common for POSTing to an endpoint, that consumes JSON.
+func (p *InfluxDBPlugin) postHelper(endpoint string, bodyBuffer *bytes.Buffer) error {
```

Comment:
Good question!

I'd flip it the other way, though - in general, it's better to think of `io.Reader` as the default parameter type for input that we only plan to read from, unless we have a strong reason to use something else instead. It makes it a lot easier to chain functions together, since that's the way a lot of the standard library (and other libraries) are intended to be used. (Case-in-point: the function we ultimately pass the buffer into accepts an `io.Reader`!)

This is a simple case, but you can see in other tests for veneur, where we want to test mocked compressed responses stored in a file, we can just do:

```go
     f, _ := os.Open("fixtures/request.json")
     gz, _  := gzip.NewReader(f)
     functionIWantToTest(gz)
```

without having to read the entire file, only to pass the contents back into a `bytes.Buffer` again.

I happened to pick two examples of readers which return errors on initialization, but for readers which don't (like `bytes.Buffer`, etc.), you can chain them all inline.


In general, the `bytes.Buffer` family is intended to be used as an easy way to create Readers and Writers, so using it as a parameter type is a mild anti-pattern - not always wrong, but usually not the cleanest answer. 


In fact, if we wanted to go all-in just to prove a point, we could even do `var buff io.ReadWriter = &bytes.Buffer{}`, and the remaining code in that function will work as-is! That's another sign that we really just want the interface type, not the underlying concrete type.